### PR TITLE
Implement input-key-down event and update the autosuggest example

### PR DESCRIPTION
**Pull request by buihoanghai**
Related to this FR https://github.com/ampproject/amphtml/issues/12518

### CODE REVIEW

**Comment by aghassemi** [comment]
Thanks @buihoanghai, /to @cvializ for review.

File: src/service/action-impl.js [hunk 1]
```diff
@@ -202,6 +202,7 @@ export class ActionService {
     this.addEvent('change');
     this.addEvent('input-debounced');
     this.addEvent('input-throttled');
+    this.addEvent('input-key-down');
```
Comment (cvializ) [inline src/service/action-impl.js]: I think the event should be called `keydown` to correspond to the HTML event.

File: examples/autosuggest.amp.html [hunk 2]
```diff
@@ -85,7 +94,12 @@ <h1>Programming language search</h1>
             autosuggest: event.value
           }),
           autosuggest-list.show,
-          results.hide"
+          results.hide;
+          input-key-down:AMP.setState({
+            index: event.keyCode == 40 ? (index < data.total - 1 ? index + 1 : 0) :
+              event.keyCode == 38 ? (index > 0 ? index - 1 : data.total - 1) : 0,
+              show: true
```
Comment (cvializ) [inline examples/autosuggest.amp.html]: This bind variable is unused.

File: examples/autosuggest.amp.html [hunk 3]
```diff
@@ -95,10 +109,11 @@ <h1>Programming language search</h1>
       layout="fixed-height"
       height="120"
       src="/form/autosuggest/query"
-      [src]="'/form/autosuggest/query?q=' + (autosuggest || '')"
+      [state]="data.items"
       id="autosuggest-list"
       hidden
     >
+      <span [text]="data.total"></span>
```
Comment (cvializ) [inline examples/autosuggest.amp.html]: This is just a number without any description. Was it a leftover debug output?

File: examples/autosuggest.amp.html [hunk 4]
```diff
@@ -65,6 +65,15 @@
   </style>
 </head>
 <body>
+<amp-state id="data"
+           src="/form/autosuggest/query"
+           [src]="'/form/autosuggest/query?q=' + (autosuggest || '')">
+  <script type="application/json">
```
Comment (cvializ) [inline examples/autosuggest.amp.html]: The json script tag is unnecessary if you have a src.

**Comment by buihoanghai** [comment]
Thanks @cvializ

File: src/service/action-impl.js [hunk 5]
```diff
@@ -202,6 +202,7 @@ export class ActionService {
     this.addEvent('change');
     this.addEvent('input-debounced');
     this.addEvent('input-throttled');
+    this.addEvent('input-keydown');
```
Comment (cvializ) [inline src/service/action-impl.js]: sorry, I meant replace `input-keydown` with `keydown`

File: examples/autosuggest.amp.html [hunk 6]
```diff
@@ -112,11 +120,12 @@ <h1>Programming language search</h1>
           {{#results}}
             <div
               class="select-option no-outline"
+              [class]="'select-option no-outline '.concat(index == {{index}} ? 'active' : '')"
```
Comment (cvializ) [inline examples/autosuggest.amp.html]: nit: just use `+` instead of `concat`

**Comment by cvializ** [comment]
Thanks for making those changes. However, I encounter some unexpected behavior when I use the demo.

When I select "Asp" with the keyboard and press the "enter" key, the submitted query is "a", but it should be "Asp".

out

You could try adding an additional `keydown` event binding that changes the value of the input and then submits the form, or a different approach if you feel that would be better.

**Comment by cvializ** [comment]
Thanks! I now see the keyboard selection working correctly. However, now when I click on a result, the form does not submit the tapped result.

**Comment by buihoanghai** [comment]
Sorry for making you review many times

**Comment by cvializ** [comment]
It's no problem, I appreciate your continued effort as you and I debug this : ) I am still seeing some misbehavior on clicking results.

out

**Comment by buihoanghai** [comment]
Hi @cvializ  Seemly this issue come from amp-selector. When I try hard click on the result( show background focus) It will work fine. If normal click it wont.

File: examples/autosuggest.amp.html [hunk 7]
```diff
@@ -53,19 +53,48 @@
...
+            query: event.value,
+            autosuggest: event.value
+          }),
+          autosuggest-list.show,
+          results.hide;
+          change: AMP.setState({
+            query: index >=0 ? data.items[0].results[index].name : event.value,
+            autosuggest: index >=0 ? data.items[0].results[index].name : event.value
+          }),search-form.submit;
+          keydown:AMP.setState({
```
Comment (cathyxz) [inline examples/autosuggest.amp.html]: https://github.com/ampproject/amphtml/issues/10947 would probably be a cleaner way to implement this. We could potentially do something like this: 
```
    this.registerAction('navigate', this.navigationKeyDownHandler_.bind(this));
```

and then do `keydown:my-selector-id.navigate` here, and I believe you wouldn't need all the index changes.

**Comment by cathyxz** [comment]
Also, you're on a version of our build system where `gulp` doesn't work properly and you have to run `gulp build` before `gulp` every time you make changes... Suggest rebasing to master for the build changes. =)

**Comment by cathyxz** [comment]
Ping @buihoanghai

**Comment by buihoanghai** [comment]
Sorry @cathyxz I just back from vacation. Will take care soon :)

**Comment by buihoanghai** [comment]
@cathyxz I addressed your comment. Pls check it again

File: extensions/amp-selector/0.1/amp-selector.js [hunk 8]
```diff
@@ -436,7 +440,7 @@ export class AmpSelector extends AMP.BaseElement {
         return;
     }
 
-    event.preventDefault();
+    //event.preventDefault();
```
Comment (cathyxz) [inline extensions/amp-selector/0.1/amp-selector.js]: Why do we need to take this out?

File: extensions/amp-selector/0.1/amp-selector.js [hunk 9]
```diff
@@ -135,6 +135,10 @@ export class AmpSelector extends AMP.BaseElement {
         this.toggle_(args['index'], args['value']);
       }
     }, ActionTrust.LOW);
+
+    this.registerAction('invocate', invocation => {
+      this.keyDownHandler_(invocation['event']);
```
Comment (cathyxz) [inline extensions/amp-selector/0.1/amp-selector.js]: `invocation.event`

File: extensions/amp-selector/0.1/amp-selector.js [hunk 10]
```diff
@@ -467,11 +471,14 @@ export class AmpSelector extends AMP.BaseElement {
   selectionKeyDownHandler_(event) {
     const keyCode = event.keyCode;
     if (keyCode == KeyCodes.SPACE || keyCode == KeyCodes.ENTER) {
+      let target = event.target;
       if (this.options_.includes(event.target)) {
         event.preventDefault();
-        const el = dev().assertElement(event.target);
-        this.onOptionPicked_(el);
+      } else {
```
Comment (cathyxz) [inline extensions/amp-selector/0.1/amp-selector.js]: I totally understand why this is reasonable for the autosuggest, but does this seem like a reasonable update to `<amp-selector>` globally? Same question re the `invocate` action registered above. 

/to @aghassemi @cvializ

Comment (cathyxz) [reply extensions/amp-selector/0.1/amp-selector.js]: Nvm. This is past me code and also past me being dumb. Let me check this again...

File: extensions/amp-selector/0.1/amp-selector.js (hunk 8 above)
Comment (cathyxz) [reply extensions/amp-selector/0.1/amp-selector.js]: Actually nvm. We can delete it instead of commenting it out.

**Comment by cvializ** [comment]
This PR has been stale for some time, and some of its functionality has been implemented by other PRs. I will close this one for now. If you are still interested in merging it, comment and we can reopen it for review.

**Pull request closed without merging**